src/Components/test/testassets/BasicTestApp/FormsTest/TypicalValidationComponent.cshtml

```diff
@@ -0,0 +1,83 @@
+@using System.ComponentModel.DataAnnotations
+@using Microsoft.AspNetCore.Components.Forms
+
+<EditForm Model="@person" OnValidSubmit="@HandleValidSubmitAsync">
+    <DataAnnotationsValidator />
+
+    <p>
+        Name: <InputText bind-Value="@person.Name" ValueExpression="@(() => person.Name)" />
+    </p>
+    <p>
+        Age (years): <InputNumber bind-Value="@person.AgeInYears" ValueExpression="@(() => person.AgeInYears)" />
+    </p>
+    <p>
+        Height (optional): <InputNumber bind-Value="@person.OptionalHeight" ValueExpression="@(() => person.OptionalHeight)" />
+    </p>
+    <p>
+        Description: <InputTextArea bind-Value="@person.Description" ValueExpression="@(() => person.Description)" />
+    </p>
+    <p>
+        Renewal date: <InputDate bind-Value="@person.RenewalDate" ValueExpression="@(() => person.RenewalDate)" />
+    </p>
+    <p>
+        Expiry date (optional): <InputDate bind-Value="@person.OptionalExpiryDate" ValueExpression="@(() => person.OptionalExpiryDate)" />
+    </p>
+    <p>
+        Ticket class:
+        <InputSelect bind-Value="@person.TicketClass" ValueExpression="@(() => person.TicketClass)">
+            <option>(select)</option>
+            <option value="@TicketClass.Economy">Economy class</option>
+            <option value="@TicketClass.Premium">Premium class</option>
+            <option value="@TicketClass.First">First class</option>
+        </InputSelect>
+    </p>
+    <p>
+        Accepts terms: <InputCheckbox bind-Value="@person.AcceptsTerms" ValueExpression="@(() => person.AcceptsTerms)" />
+    </p>
+
+    <button type="submit">Submit</button>
+
+    <ValidationSummary />
+</EditForm>
+
+<ul>@foreach (var entry in submissionLog) { <li>@entry</li> }</ul>
+
+@functions {
+    Person person = new Person();
+
+    // Usually this would be in a different file
+    class Person
+    {
+        [Required, StringLength(10)]
+        public string Name { get; set; }
+
+        [Range(0, 200)]
+        public int AgeInYears { get; set; }
+
+        public float? OptionalHeight { get; set; }
+
+        public DateTime RenewalDate { get; set; } = DateTime.Now;
+
+        public DateTimeOffset? OptionalExpiryDate { get; set; }
+
+        [Required, Range(typeof(bool), "true", "true", ErrorMessage = "Must accept terms")]
+        public bool AcceptsTerms { get; set; }
+
+        [Required, StringLength(20)]
+        public string Description { get; set; }
+
+        [Required, EnumDataType(typeof(TicketClass))]
+        public TicketClass TicketClass { get; set; }
+    }
+
+    enum TicketClass { Economy, Premium, First }
+
+    List<string> submissionLog = new List<string>(); // So we can assert about the callbacks
+
+    Task HandleValidSubmitAsync(EditContext editContext)
+    {
+        submissionLog.Add("OnValidSubmit");
+        StateHasChanged(); // Temporary workaround for event routing bug
+        return Task.CompletedTask;
+    }
+}
```

Comment:
I'm really really happy with how this is going to look once the compiler support is merged. This is what scaffolding is going to generate, it's great to see it come together.